module/zfs/zio.c

```diff
@@ -1773,10 +1773,6 @@ zio_write_compress(zio_t *zio)
 
 		if (pass >= zfs_sync_pass_dont_compress)
 			compress = ZIO_COMPRESS_OFF;
-
-		/* Make sure someone doesn't change their mind on overwrites */
-		ASSERT(BP_IS_EMBEDDED(bp) || MIN(zp->zp_copies + BP_IS_GANG(bp),
-		    spa_max_replication(spa)) == BP_GET_NDVAS(bp));
```

Comment:
I think it would still be valid to keep an assertion here but maybe it can be simplified so that we don't have to deal with gang logic that may change in the future. We could instead write the assertion as:
```
ASSERT(BP_IS_EMBEDDED(bp) || BP_IS_GANG(bp) || MIN(zp->zp_copies, spa_max_replication(spa)) == BP_GET_NDVAS(bp));
```
This would still provide protection for the common case to ensure that we don't modify the copies when doing overwrites. Also updating the comment to reflect that overwriting gang blocks are ignored here since their logic is dictated by `zio_write_gang_block`. Just a thought.

Reply:
Updated

Reply:
I don't think that this assertion is protecting anything critical.  If we do try to write the same block in the same txg with a different number of copies, we might end up using the first-requested number of copies.

The assertion confirms our understanding of how the code works, but it needs to be kept in sync with logic that's distributed in several different places in the code, leading to it getting out of sync and causing this bug.

So I think it would be preferable to remove it.  But, I'm fine with keeping it in as an assertion, disabled on production builds, so it has negligible impact if it gets out of sync again.